examples/rcn.py

```diff
@@ -0,0 +1,392 @@
+    for i, pc in enumerate(pooled_channels):
+        for channel_offset, factor in pooled_channel_weights:
+            ch = (i + channel_offset) % num_orients
+            pos_chan = localized[ch]
+            if factor != 1:
+                pos_chan[pos_chan > 0] *= factor
+            np.maximum(pc, pos_chan, pc)
+
+    bu_msg = np.array(pooled_channels)
+    bu_msg[bu_msg == 0] = -1
+    return bu_msg
+
+
+# %% [markdown]
+# ## 5.1.1 Visualizing bu_msg for a sample image
+
+# %% [markdown]
+# bu_msg has shape (16, H, W) where each 1 <= f <= 16 denotes the present or absense of a oriented edge
+
+# %%
+r_test_img = test_set[4]
+r_bu_msg = get_bu_msg(r_test_img)
+img = np.ones((200, 200))
+fig, ax = plt.subplots(1, 2, figsize=(20, 10))
+ax[0].imshow(r_test_img, cmap="gray")
+ax[0].axis("off")
+ax[0].set_title("Input image", fontsize=30)
+for i in range(r_bu_msg.shape[0]):
+    img[r_bu_msg[i] > 0] = 0
+
+ax[1].imshow(img, cmap="gray")
+ax[1].axis("off")
+ax[1].set_title("Max filter response across 16 channels", fontsize=30)
+fig.tight_layout()
+
+
+# %% [markdown]
+# ## 5.2 Run map product inference on all test images
+
+# %%
+def get_evidence(bu_msg: np.ndarray, frc: np.ndarray):
+    """Function to get evidence
+
+    Args:
+        bu_msg: Array of shape (n_features, 200, 200). Contains BU messages
+        frc: Array of shape (n_frcs, 3).
+
+    Returns:
+        evidence; Array of shape (n_frcs, M). Contains evidence
+    """
+    evidence = np.zeros((frc.shape[0], M))
+    for v, (f, r, c) in enumerate(frc):
+        evidence[v] = bu_msg[f, r - hps : r + hps + 1, c - vps : c + vps + 1].ravel()
+
+    return evidence
+
+
+# %%
+frcs_dict = {model_idx: frcs[model_idx] for model_idx in range(frcs.shape[0])}
+run_bp, _, get_beliefs = graph.BP(fg.bp_state, 30)
+scores = np.zeros((len(test_set), frcs.shape[0]))
+map_states_dict = {}
+
+for test_idx in range(len(test_set)):
+    img = test_set[test_idx]
+
+    start = time.time()
+    bu_msg = get_bu_msg(img)
+    evidence_updates = jax.tree_util.tree_map(
+        lambda frc: get_evidence(bu_msg, frc), frcs_dict
+    )
+    end = time.time()
+    print(f"Initializing evidences took {end-start:.3f} seconds for image {test_idx}.")
+
+    start = end
+    map_states = graph.decode_map_states(
+        get_beliefs(run_bp(evidence_updates=evidence_updates))
+    )
+    end = time.time()
+    print(f"Max product inference took {end-start:.3f} seconds for image {test_idx}.")
+
+    map_states_dict[test_idx] = map_states
+    start = end
+    score = tree_util.tree_multimap(
+        lambda evidence, map: jnp.sum(evidence[jnp.arange(map.shape[0]), map]),
+        evidence_updates,
+        map_states,
+    )
+    for ii in score:
+        scores[test_idx, ii] = score[ii]
+    end = time.time()
+    print(f"Computing scores took {end-start:.3f} seconds for image {test_idx}.")
+
+
+# %% [markdown]
+# # 6. Compute metrics (accuracy)
+
+# %%
+best_model_idx = np.argmax(scores, axis=1)
```

Comment:
what if we have two classes that score the same?

Reply:
`argmax` will chose the one with smaller value.

Reply:
I know but that is wrongly reducing the accuracy.

Reply:
If two templates have the same score, I am not sure what we can do in this case. It amounts to an arbitrarily selecting between the two (or more) of them.